src/pkcs11.cpp

```diff
@@ -0,0 +1,288 @@
+   << " Line: " << __LINE__ << endl; \
+   return -1; })
+
+/* HSM default information */
+static const struct lmp_hsm {
+	string token;
+	string tls_lbl;
+	string tls_id;
+	string crt_lbl;
+	unsigned char crt_id;
+} hsm_cfg = {
+	.token = HSM_TOKEN_STR,
+	.tls_lbl = HSM_TLS_STR,
+	.tls_id = HSM_TLS_ID_STR,
+	.crt_lbl = HSM_CRT_STR,
+	.crt_id = HSM_CRT_ID,
+};
+
+int pkcs11_get_uuid(lmp_options &opt)
+{
+	PKCS11_SLOT *slots{nullptr};
+	PKCS11_CTX *ctx{nullptr};
+	unsigned int nslots = 0;
+
+	if (opt.hsm_module.empty())
+		return 0;
+
+	ctx = PKCS11_CTX_new();
+	if (!ctx)
+		leave;
+
+	if (PKCS11_CTX_load(ctx, opt.hsm_module.c_str()))
+		leave;
+
+	if (PKCS11_enumerate_slots(ctx, &slots, &nslots))
+		leave;
+
+	/* UUID format is a requirement for HSM modules (mainly OP-TEE) */
+	std::regex UUID("("
+			"[a-f0-9]{8}-"
+			"[a-f0-9]{4}-"
+			"[a-f0-9]{4}-"
+			"[a-f0-9]{4}-"
+			"[a-f0-9]{12}"
+			")");
+	std::smatch match;
+
+	for (int i = 0; i < nslots && opt.uuid.empty(); i++) {
+		string slot_info = string(slots[i].description);
+
+		std::regex_search(slot_info, match, UUID);
+		if (match.size() > 0)
+			opt.uuid = match.str(1);
+	}
+
+	PKCS11_release_all_slots(ctx, slots, nslots);
+	PKCS11_CTX_unload(ctx);
+	PKCS11_CTX_free(ctx);
+
+	return 0;
+}
+
+/* Initialize HSM_TOKEN_STR (aktualizr token) */
+static int pkcs11_initialize(const lmp_options &opt)
+{
+	PKCS11_TOKEN *token = NULL;
+	PKCS11_SLOT *slots = NULL;
+	PKCS11_SLOT *slot = NULL;
+	PKCS11_CTX *ctx = NULL;
+	unsigned int nslots;
+	char label[32] = { ' ' };
+
+	ctx = PKCS11_CTX_new();
+
+	if (PKCS11_CTX_load(ctx, opt.hsm_module.c_str()))
+		leave;
+
+	if (PKCS11_enumerate_slots(ctx, &slots, &nslots))
+		leave;
+
+	slot = PKCS11_find_token(ctx, slots, nslots);
+
+	while (slot && slot->token && slot->token->initialized)
+		slot = PKCS11_find_next_token(ctx, slots, nslots, slot);
+
+	if (!slot || !slot->token || slot->token->initialized)
+		leave;
+
+	/* We have an un-initialized token */
+	token = slot->token;
+
+	/* Remove the 'end of string' character from the label */
+	memcpy(label, HSM_TOKEN_STR, strlen(HSM_TOKEN_STR));
+
+	if (PKCS11_init_token(token, opt.hsm_so_pin.c_str(),
+			      (const char *)label))
+		leave;
+
+	if (PKCS11_open_session(slot, 1))
+		leave;
+
+	if (PKCS11_login(slot, 1, opt.hsm_so_pin.c_str()))
+		leave;
+
+	if (PKCS11_init_pin(token, opt.hsm_pin.c_str()))
+		leave;
+
+	if (PKCS11_logout(slot))
+		leave;
+
+	PKCS11_release_all_slots(ctx, slots, nslots);
+	PKCS11_CTX_unload(ctx);
+	PKCS11_CTX_free(ctx);
+
+	cout << "PKCS11 token " << HSM_TOKEN_STR << " initialized" << endl;
+
+	return 0;
+}
+
+/* Create an EC keypair in the HSM (not TPM) and generate a CSR */
+int pkcs11_create_csr(const lmp_options &opt, string &pkey, string &csr)
+{
+	PKCS11_SLOT *slots = NULL;
+	PKCS11_SLOT *slot = NULL;
+	PKCS11_CTX *ctx = NULL;
+	PKCS11_KEY *key = NULL;
+	EVP_PKEY *prv = NULL;
+	EVP_PKEY *pub = NULL;
+	unsigned int nslots = 0;
+	unsigned int n = 0;
+	PKCS11_EC_KGEN ec = {
```

Comment:
I get the following compilation error. Does this code depend on a specific libp11 version?

```
/work/fio/prjs/lmp-device-register/src/pkcs11.cpp: In function 'int pkcs11_create_csr(const lmp_options&, std::string&, std::string&)':
/work/fio/prjs/lmp-device-register/src/pkcs11.cpp:143:9: error: 'PKCS11_EC_KGEN' was not declared in this scope; did you mean 'PKCS11_TOKEN'?
  143 |         PKCS11_EC_KGEN ec = {
      |         ^~~~~~~~~~~~~~
      |         PKCS11_TOKEN

```

Reply:
yes, please see this: https://github.com/foundriesio/meta-lmp/pull/1104/commits/804c6767fe50a60755e4e20e976d61f0ea6b3b96

Reply:
I see the same compilations errors:
```
error: ‘PKCS11_EC_KGEN’ was not declared in this scope; did you mean ‘PKCS11_TOKEN’?
error: ‘PKCS11_KGEN_ATTRS’ was not declared in this scope; did you mean ‘PKCS11_KEYGEN_FAILED’?
error: ‘attr’ was not declared in this scope
error: ‘ec’ was not declared in this scope
```

Reply:
it integrates this better documented patch https://github.com/foundriesio/lmp-device-register/commit/ed399779f0576a0ad08c71e5aec8f055a202410b

Reply:
I can build the master branch that integrates that patch lmp-dev-reg on a local host (no HSM support) while I cannot build the given PR.

Reply:
It looks like this commit https://github.com/foundriesio/meta-lmp/commit/804c6767fe50a60755e4e20e976d61f0ea6b3b96 is not merged into meta-lmp's main branch. How can I build this version of lmp-dev-reg then.

Reply:
Maybe we should introduce some build flag so the patch is not required for non-lmp or non-HSM usage.
Ideally, it would be great to support softhsm without that patch if it's possible.

Reply:
> Maybe we should introduce some build flag so the patch is not required for non-lmp or non-HSM usage. Ideally, it would be great to support softhsm without that patch if it's possible.

`-DDISABLE_PKCS11=ON ` to disable building-in pkcs11 support

Reply:
It works, thanks.